### PR TITLE
Add caching to FirestoreUuidTable

### PR DESCRIPTION
**Pull request by as2388**
This speeds up repeated requests to the same uuid table instance and, more importantly, drastically reduces the number of Firestore reads needed. Testing on JPLG-2020, this reduces the number of documents read from Firestore from 347220 per run of fetch raw data to 23148, with the only change needed in the pipeline being a version number bump in the Pipfile.

### CODE REVIEW

File: id_infrastructure/firestore_uuid_table.py [hunk 1]
```diff
@@ -119,11 +134,18 @@ def uuid_to_data(self, uuid_to_lookup):
         raise LookupError() 
 
     def uuid_to_data_batch(self, uuids_to_lookup):
+        # Serve the request from the cache if possible, saving network request time + Firestore read costs
+        uuids_to_lookup = set(uuids_to_lookup)
+        if len(uuids_to_lookup - set(self._mappings_cache.values())) == 0:
```
Comment (IsaackMwenda) [inline id_infrastructure/firestore_uuid_table.py:139]: if we are partly missing some uids e.g if running the pipeline twice concurrently(which is rare) we should probably return the uuids available in cache and only look up the rest from firestore.

Comment (as2388) [reply id_infrastructure/firestore_uuid_table.py:139]: As far as I could tell from the documentation, there isn't an efficient way to look up a list of ids from Firestore. The best I could find is 'where in' but that only supports 10 at a time so we might need to make thousands of round trips, which I thought would be very slow.

File: id_infrastructure/firestore_uuid_table.py [hunk 2]
```diff
@@ -21,17 +21,27 @@ def __init__(self, table_name, crypto_token_path, uuid_prefix):
...
+        # overwriting any existing data just in case it's not up-to-date, which likely means the table was being used
+        # concurrently)
+        if len(self._mappings_cache) == 0:
+            log.info(f"Sourcing uuids for {len(list_of_data_requested)} data items from Firestore...")
+            existing_mappings = dict()
+            for mapping in self._client.collection(f"tables/{self._table_name}/mappings").get():
+                existing_mappings[mapping.id] = mapping.get(_UUID_KEY_NAME)
+        else:
+            log.info(f"Sourcing uuids for {len(list_of_data_requested)} data items from cache...")
+            existing_mappings = dict(self._mappings_cache)
```
[the remark this answers is not in the document]
Comment (as2388) [reply id_infrastructure/firestore_uuid_table.py:44]: Yeah  was trying to make a copy here. Let me check if that is necessary and make that intent clearer (with .copy) if it is.

[the remark this answers is not in the document]
Comment (as2388) [reply id_infrastructure/firestore_uuid_table.py]: Changed to .copy()

File: id_infrastructure/firestore_uuid_table.py (hunk 1 above)
Comment (IsaackMwenda) [reply id_infrastructure/firestore_uuid_table.py:139]: Okay, thanks

**Pull request merged**